src/System.Private.CoreLib/shared/System/Array.cs

```diff
@@ -1132,23 +1232,84 @@ public static int LastIndexOf(Array array, object? value, int startIndex, int co
+                    if (TryGenericLastIndexOf<char>(array, value, startIndex, count, out result)) return result;
+                    break;
+                case TypeCode.Double:
+                    if (TryGenericLastIndexOf<double>(array, value, startIndex, count, out result)) return result;
+                    break;
+                case TypeCode.Int16:
+                    if (TryGenericLastIndexOf<short>(array, value, startIndex, count, out result)) return result;
+                    break;
+                case TypeCode.Int32:
+                    if (TryGenericLastIndexOf<int>(array, value, startIndex, count, out result)) return result;
+                    break;
+                case TypeCode.Int64:
+                    if (TryGenericLastIndexOf<long>(array, value, startIndex, count, out result)) return result;
+                    break;
+                case TypeCode.SByte:
+                    if (TryGenericLastIndexOf<sbyte>(array, value, startIndex, count, out result)) return result;
+                    break;
+                case TypeCode.Single:
+                    if (TryGenericLastIndexOf<float>(array, value, startIndex, count, out result)) return result;
+                    break;
+                case TypeCode.UInt16:
+                    if (TryGenericLastIndexOf<ushort>(array, value, startIndex, count, out result)) return result;
+                    break;
+                case TypeCode.UInt32:
+                    if (TryGenericLastIndexOf<uint>(array, value, startIndex, count, out result)) return result;
+                    break;
+                case TypeCode.UInt64:
+                    if (TryGenericLastIndexOf<ulong>(array, value, startIndex, count, out result)) return result;
+                    break;
+            }
+
+            static bool TryGenericLastIndexOf<T>(Array array, object? value, int startIndex, int count, out int result) where T : struct
+            {
+                if (array is T[] arrayOfT)
                 {
-                    object? obj = array.GetValue(i);
-                    if (obj == null)
+                    if (value is null)
                     {
-                        if (value == null)
-                            return i;
+                        result = -1;
+                        return true;
                     }
-                    else
+
+                    if (value is T valueOfT)
                     {
-                        if (obj.Equals(value))
-                            return i;
+                        result = LastIndexOf<T>(arrayOfT, valueOfT, startIndex, count);
+                        return true;
                     }
                 }
+
+                result = default;
+                return false;
+            }
+
+            for (int i = startIndex; i >= endIndex; i--)
+            {
+                object? obj = array.GetValue(i);
+                if (obj == null)
+                {
+                    if (value == null)
```

Comment:
I see a lot of `is null` in new code and here `== null` was added. Does the jit care which in this case? Is there a type or performance consideration to push in either direction?

Reply:
It matters only in the few rare cases where `==` operator is overload. There is no difference (the IL is exactly same) in 99% of cases.